linux-pipewire: check pipewire buffers for corruption

columbarius opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->

This patch checks the buffer received from PipeWire for flags indicating that the buffer is corrupt.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
Portals can set a buffer as corrupted via flags. This allows obs to skip those frames and avoid trying to import them.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Modified version of [xdg-desktop-portal-wlr](https://github.com/emersion/xdg-desktop-portal-wlr) to alternate between those flags for a given time duration. With this patch your screencast will freeze until a new frame without a flag arrives. Mouse pointer should still be updated when it is not embedded. 


### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
- Bug fix (non-breaking change which fixes an issue)
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

kkartaltepe commented on plugins/linux-capture/pipewire.c:
```diff
@@ -472,6 +488,11 @@ static void on_process_cb(void *user_data)
 		}
+		if ((buffer->datas[0].chunk->flags & SPA_CHUNK_FLAG_CORRUPTED) >
+		    0) {
+			blog(LOG_ERROR, "[pipewire] buffer plane corrupt");
```
Should this be log_debug like the other reports?

columbarius replied:
Oh thanks, missed to change that back.

kkartaltepe commented:
Is this supposed to solve #5468 ?

columbarius commented:
> Is this supposed to solve #5468 ?

No

GeorgesStavracas reviewed (changes requested):
Commit message typo: `metadate` → `metadata`

Could you expand a bit more the context around these CORRUPTED flags? The documentation I found is painfully small, and I don't think I understand when nor how this flag is used. Is it PipeWire itself that sets it? The compositor? Is skipping the frame the best strategy to handle it?

columbarius commented:
> Could you expand a bit more the context around these CORRUPTED flags? The documentation I found is painfully small, and I don't think I understand when nor how this flag is used. Is it PipeWire itself that sets it? The compositor? Is skipping the frame the best strategy to handle it?

This flag is set by the producer, in our case the compositor. Afaik this flag can be used to mark "broken" buffers, without specifying what the issue is. xdpw will use that flag if a buffer copy by the compositor failed. Since we manage the buffer the memory won't be invalid but the content might be corrupt (filled with garbage). The v4l2 plugin [1] uses this flag for the same purpose. Since a buffer with this flag won't contain wanted contend I can't think of a better solution than skipping it.  In my testing obs just reused the last imported texture and continued after a non corrupted frame was received.

[1] https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/plugins/v4l2/v4l2-utils.c#L1264

columbarius commented:
Updated this to incorporate both meta and junk flags and added a revertible commit for current GNOME behaviour.

The hinted documentation for ignoring the size in case of DMA-BUFs will be written by me probably this week, but got an ack from mesa devs. Will link the PipeWire MR here.

columbarius commented:
> The documentation I found is painfully small

Try to extend that at the same time.

columbarius commented:
https://gitlab.freedesktop.org/pipewire/pipewire/-/merge_requests/1189

GeorgesStavracas reviewed (changes requested):
Two comments, but this looks generally good to me.

GeorgesStavracas commented:
I forgot to mention, but could you please rebase against the main branch, and remove the merge commits?

columbarius commented on plugins/linux-pipewire/pipewire.c:
```diff
@@ -480,7 +481,15 @@ static void on_process_cb(void *user_data)
 
 	buffer = b->buffer;
-	has_buffer = buffer->datas[0].chunk->size != 0;
+	if ((header = spa_buffer_find_meta_data(buffer, SPA_META_Header,
+						sizeof(*header)))) {
+		if ((header->flags & SPA_META_HEADER_FLAG_CORRUPTED) > 0) {
+			blog(LOG_ERROR, "[pipewire] buffer is corrupt");
+			goto queue_buffer;
+		}
+	} else {
+		has_buffer = buffer->datas[0].chunk->size != 0;
+	}
```
we can't do the `has_buffer` assignment. DMABUFS are allowed to have a `chunk->size = 0` and be valid. The reason is that you don't need the size of a DMABUF to import it and gbm has no api to query it. You could do an fseek on the fd, but that would not be useful in handling the buffer anyways, as such PipeWire just allows to omit this value (set it to 0, so clients which mmap the fd won't map anything bad).

columbarius replied:
I'd rather remove the size check at all, but we need it for backwards compatibility for now.

GeorgesStavracas replied:
Hm, alright then

columbarius commented:
> I forgot to mention, but could you please rebase against the main branch, and remove the merge commits?

It's gone. Probably didn't changed the strategy when using githubs update feature.

GeorgesStavracas commented on plugins/linux-pipewire/pipewire.c:
```diff
@@ -616,6 +641,7 @@ static void on_process_cb(void *user_data)
 		obs_pw->cursor.y = cursor->position.y;
 	}
 
+queue_buffer:
```
This label is unused now

columbarius replied:
Nice catch. Thanks!

GeorgesStavracas reviewed (changes requested):
I understand why you split this in 2 separate commits - and it actually helped me review it - but I'd prefer if you could squash the second commit into the first one before merge. Please mention the same GNOME pre 43 behavior in the first commit's message. (The reason I prefer squashing is to not have any point in the commit history where the capture won't work with a subset of consumers.)

columbarius commented:
> I understand why you split this in 2 separate commits - and it actually helped me review it - but I'd prefer if you could squash the second commit into the first one before merge. Please mention the same GNOME pre 43 behavior in the first commit's message. (The reason I prefer squashing is to not have any point in the commit history where the capture won't work with a subset of consumers.)

Squashed the commit and added the note to the first commit message. Additionally I marked the workaround in code with a comment, such that it will be recognized as such and can be easily removed later.

I'll continue to use multiple commits to ease development and review, but I'm totally fine to squash them just before merging.

GeorgesStavracas reviewed (approved):
One last nit: please use `linux-pipewire:` as the commit prefix; and please start the tile in upper case. In other words, this: `linux-pipewire: Reject invalid buffers`

columbarius commented:
> One last nit: please use `linux-pipewire:` as the commit prefix; and please start the tile in upper case. In other words, this: `linux-pipewire: Reject invalid buffers`

Done ... `linux-capture`... what a nod from the past ^^`

GeorgesStavracas commented:
> Done ... linux-capture... what a nod from the past ^^`
Heh the good, old, simpler times :slightly_smiling_face:

jp9000 commented:
This change spams my log with corruption messages even though it seems to be capturing fine. Fedora 37

columbarius commented:
@jp9000 is it `[pipewire] buffer is corrupt` or `[pipewire] buffer contains corrupted data` ?
The second one should also appear when the buffer only contains an cursor update without new screen data. Maybe we should change the loglevel or message.

columbarius commented:
@GeorgesStavracas thanks for merging!

jp9000 commented:
@columbarius It's `[pipewire] buffer contains corrupted data`, the one on line 533. I just noticed it happens if I have OBS on a monitor different from the one being captured. If OBS is capturing the same monitor it's fine, but when I move it onto my other monitor, it starts spamming that in the logs in spite of seemingly capturing fine.
